### ISSUE TITLE
[feature] - Script Hot Reload

### ISSUE THREAD

**Issue by Olyno** [SkriptLang/Skript#4313]
### Suggestion

Coding a script is cool. What could be cooler than that? Reloading the script after each save, and be able to disable or enable this feature from the config file of Skript.

### Why?

It able us to be more productive when creating a project.

### Other

[Skent](https://github.com/Olyno/skent) already implement a system for that with its file watcher feature. We could use something like that.

### Agreement

- [X] I have read the guidelines above and confirm I am following them with this suggestion.

**Comment by AyhamAl-Ali** [comment]
Useful idea indeed :+1: another useful idea would be to actually improve reloading by only updating changed code/lines and keep unchanged code/lines, this will decrease the reloading time and increase performance a lot (there was a suggestion of caching which is simply almost the same thing #3003)

**Comment by Anarchick** [comment]
I agree only if it's an option in the config.sk.

I don't need this feature in local serv , only on public server ;)

**Comment by Runakai1** [comment]
I really really like this idea. It would make things easier. 
Like we did with the send/broadcast effect. We don't need to surround our objects with "%%" anymore. It's pretty nice. 
But it could really fuck up your server. What if I had a while loop without delay? That's why making it optional is good. That'd prevent the "fucking up" lol

**Comment by LowScarlet** [comment]
nope but +1

**Comment by Runakai1** [comment]
> I am mixed, I ask myself several questions:
> who will receive the script errors?
> if it's the console, some skript developers don't have access to it
> and if it's all OP players, that's not good because on many servers the OP players don't code, they are just the owners

It'll probably just as if you did /Skript reload X. The player would receive it. It's also not an endless loop since it'll only loop once the file is edited

**Comment by JonesJugHead** [comment]
> 
> 
> > I am mixed, I ask myself several questions:
> > who will receive the script errors?
> > if it's the console, some skript developers don't have access to it
> > and if it's all OP players, that's not good because on many servers the OP players don't code, they are just the owners
> 
> It'll probably just as if you did /Skript reload X. The player would receive it. It's also not an endless loop since it'll only loop once the file is edited

So the scripts would reload themselves but not indicate the errors?
I think there should be a system where skript detects which player is working on which file, for example if I do "/sk reload file.sk" and then reload the file it understands that it is me working on it and sends me the message that the file has been reloaded!

**Comment by Runakai1** [comment]
> > > I am mixed, I ask myself several questions:
> > > who will receive the script errors?
> > > if it's the console, some skript developers don't have access to it
> > > and if it's all OP players, that's not good because on many servers the OP players don't code, they are just the owners
> > 
> > 
> > It'll probably just as if you did /Skript reload X. The player would receive it. It's also not an endless loop since it'll only loop once the file is edited
> 
> So the scripts would reload themselves but not indicate the errors?
> I think there should be a system where skript detects which player is working on which file, for example if I do "/sk reload file.sk" and then reload the file it understands that it is me working on it and sends me the message that the file has been reloaded!

You're right. How is the auto reload supposed to know whom to send the errors to? Op should be fine though. You shouldn't give anyone besides yourself op rights anyway. Or just a specific permission is required

**Comment by bluelhf** [comment]
In my opinion, hot-reload is not viable because Skript operates mostly on the main thread:
1. Skript has hot-reload enabled
2. Player types the following:
  ```hs
  set {_i} to 5
  while true:
    add 1 to {_i}
```
3. Script hot-reloads
4. Server crashes
5. Player types the rest of the script:
  ```hs
  set {_i} to 5
  while true:
    add 1 to {_i}
    stop if {_i} > 100
```
6. Too late, the server already crashed

It would need to have enough of a cooldown that users have the time to write everything they want to write before a reload occurs, which would arguably be longer than the time taken to just reload the script...

The points about error output are also fair. It would be difficult to implement cleanly.
If we want to implement it, then I'd suggest something like this:
```hs
/skript hotswap test.sk
[Skript] Hotswap mode enabled for test.sk.
[Skript] Changes detected. Reloading in 5 seconds if no more changes occur.
[Skript] No further changes. Reloading test.sk...
[Skript] <insert test.sk> reload output
[Skript] Hotswapped test.sk
...
/skript hotswap test.sk
[Skript] Hotswap mode disabled for test.sk,
```

This would take care of the issues with error output, and the users would be aware of the possible dangers of hot-swapping code that's running on the main server thread.

**Comment by AyhamAl-Ali** [comment]
> In my opinion, hot-reload is not viable because Skript operates mostly on the main thread:

Hot reload should never be used in `on keydown` event, it should only trigger if you `save` the file with that you should get such crash issue unless someone saves too quickly \:D

**Comment by Olyno** [comment]
> In my opinion, hot-reload is not viable because Skript operates mostly on the main thread:
> 
> 1. Skript has hot-reload enabled
> 2. Player types the following:
> 
> ```haskell
> set {_i} to 5
> while true:
>   add 1 to {_i}
> ```
> 
> 1. Script hot-reloads
> 2. Server crashes
> 3. Player types the rest of the script:
> 
> ```haskell
> set {_i} to 5
> while true:
>   add 1 to {_i}
>   stop if {_i} > 100
> ```
> 
> 1. Too late, the server already crashed
> 
> It would need to have enough of a cooldown that users have the time to write everything they want to write before a reload occurs, which would arguably be longer than the time taken to just reload the script...
> 
> The points about error output are also fair. It would be difficult to implement cleanly.
> If we want to implement it, then I'd suggest something like this:
> 
> ```haskell
> /skript hotswap test.sk
> [Skript] Hotswap mode enabled for test.sk.
> [Skript] Changes detected. Reloading in 5 seconds if no more changes occur.
> [Skript] No further changes. Reloading test.sk...
> [Skript] <insert test.sk> reload output
> [Skript] Hotswapped test.sk
> ...
> /skript hotswap test.sk
> [Skript] Hotswap mode disabled for test.sk,
> ```
> 
> This would take care of the issues with error output, and the users would be aware of the possible dangers of hot-swapping code that's running on the main server thread.

You're not wrong, but i know that we can make a real hot reload thing. If we detect that kind of problems before reloading, we can just cancel the reload 🙂

**Comment by bluelhf** [comment]
> 
> 
> > In my opinion, hot-reload is not viable because Skript operates mostly on the main thread:
> 
> Hot reload should never be used in `on keydown` event, it should only trigger if you `save` the file with that you should get such crash issue unless someone saves too quickly :D

fair, i guess i'm used to IDEA's not-having-a-save-button

**Comment by bluelhf** [comment]
I've also made a [FileWatcher](https://gist.github.com/bluelhf/c200430dfe0e8aaa701b1a69c6f50d34) class before - this one deals with stuff like low-speed file uploads too. It's dependent on Apache Commons DigestUtils, though..

**Comment by MrScopes** [comment]
`/sk watch <script / directory | scripts | config | aliases | all>` something like this would be great, /sk watch with no args would show what you're watching, also /sk unwatch or /sk watch could be a toggle
